Shell autocompletion

magik6k opened the pull request:
Added basic auto completion for shell that can recognize files and programs. Tapping tab key twice will print all the possibilities

magik6k committed:
Add tab key handler to term.read

Adding this handler was necessary to allow creation of auto-complete mechanism

magik6k committed:
Add autocompletion to shell

fnuecke commented on src/main/resources/assets/opencomputers/loot/OpenOS/lib/term.lua:
```diff
@@ -265,6 +277,8 @@ function term.read(history, dobreak)
       up()
     elseif code == keyboard.keys.down then
       down()
+    elseif code == keyboard.keys.tab then
```
Maybe better `elseif code == keyboard.keys.tab and not hint then`?

magik6k replied:
Could be, but probably you meant `elseif code == keyboard.keys.tab and hint then`

fnuecke replied:
Gah, yes I did.

fnuecke commented on src/main/resources/assets/opencomputers/loot/OpenOS/bin/sh.lua:
```diff
@@ -148,6 +148,76 @@ end
+  local res = {}
+  for dir in string.gmatch(os.getenv("PATH"), "[a-zA-Z0-9/.]+") do
+    for file in fs.list(dir) do
+      if string.match("/" .. file, "/" .. pattern) and file:match("(.+)[.]lua") then
```
Hmm, why the `"/" ..`s? Or rather, why not `"^" .. pattern`? Won't this potentially match things that just _contain_ the current text (which I don't think it should)? Ditto for the `file:match`, should probably also start with a `^`, no?

magik6k replied:
Using `"^"..pattern` performs better, second match does it's work well - checks for `*.lua`. There is way to make one match, as I didn't know the `"^"`, and it's `if string.match(file, "^" .. pattern .. "(.+)[.]lua")  then`

magik6k committed:
Move hint check to onKeyDown

magik6k committed:
Made getMatchingPrograms more accurate

magik6k commented on src/main/resources/assets/opencomputers/loot/OpenOS/bin/sh.lua:
```diff
@@ -148,6 +148,76 @@ end
+    local matches = getMatchingFiles(after)
+    if #matches == 1 then
+      lastSearch = ""
+      return base .. space .. matches[1] .. " "
```
This is broken, repairing it..

magik6k committed:
Repaired file completion behavior

fnuecke commented:
This looks pretty cool, thanks :-)

I'd like to talk about a design decision, though, that I'm not sure about either way, myself: would it be better to have the 'preview' logic inside term itself? `hint` would then just return a table of 'completions'.
- Pro: more context, allows smarter behavior (e.g. caching results from `hint` on repetitive tab presses).
- Pro: centralized, less redundant code (people would just copy from sh.lua anyway).
- Con: must not print across multiple lines, so something like cycling through alternatives would have to be implemented instead (other ideas?).
- Con: less flexible.

I am currently slightly favoring putting it into term.read.

magik6k commented:
Well, the logic is hard to make useful in every use-case, for instance when you want to complete a file, when completion is a directory, it has `/` appended, else it gets a space. In fact `hint` should be able to return both string for completion and table (and/or iterator) for list of matching completions - and in case table contains only 1 element, it should also get printed, cause it would be hard to put it well into already written string. Let's say that user has written `/bin/som`, shall the completion be `/bin/somsomefile.txt` or `somefile.txt`? I guess neither of those are good. But still it's a good idea.

magik6k committed:
Hint function can now return tables and iterators

Returning table or iterator will display set of available completion options. When user will return table, he/she should also provide prompt function for (re)drawing prompt if one is used is his/her program

magik6k committed:
Fix term.lua completion list printing

magik6k committed:
Fixed the fix

magik6k committed:
Update hintHandler

Updated hintHandler to return table when tab was dual-tapped and there are multiple possible completions

magik6k commented:
Okay, `hint` may now return tables, also I added `prompt` parameter which may be function drawing prompt or just prompt string.

fnuecke commented:
Uhm, from what I see (sorry, haven't had the time to actually run it), this now prints to multiple lines in `term.read()`? I don't really think that's a good idea, since it vastly limits the use of tables as hints. It's what I meant in my earlier reply - instead I think it'd be better to just 'cycle' through the alternatives (i.e. say you have `/b`, tab -> `/bin`, tab -> `/boot`, repeat). That would also make the prompt parameter superfluous.

magik6k commented:
I haven't thought of that, and it seems good, as shell creator still may print hint list inside . I'll code that and push later today.

magik6k committed:
Changed term.read hint behaviour

When hint function returns table, It no longer writes it to screen. The result is cached, and user may view hintd by pressing tab key repetitively.

magik6k committed:
Update shell to work with new term.lua

magik6k commented:
Okay, it's now working in way you told. If you want test it grab some OC with os on hdd+internet card and do:
`pastebin run KaATrNkb`
`mpt ppa add m6koc`
`rm /lib/term.lua`
`mpt -Sy mesh`
`reboot`
`mesh`

fnuecke commented on src/main/resources/assets/opencomputers/loot/OpenOS/lib/term.lua:
```diff
@@ -109,6 +109,9 @@ function term.read(history, dobreak)
   local scrollX, scrollY = 0, #history - 1
   local cursorX = 1
 
+  local hintCache = (type(hint)=="table" and #hint > 1)and hint
```
Sorry if I was unclear about that - I never really meant `hint` itself should be a table, I think it's fine to always have that as a function. I just meant it could _return_ a table that could then be cached. I think that'd simplify the code a bit, no?

I promise I'll find time to actually test it out tomorrow, busy day today.

magik6k replied:
I understood what you mentioned above, but I thought that it may be simple addition, that can be used by someone creating simpler programs using static set of commands. It's just a simple feature.

fnuecke replied:
Ah, all right then. I'll clean this up a little, there are some things that seem odd to me (e.g. the use of `after` in `tab()` which isn't defined anywhere).

I'll also try to adapt it to take into account the cursor position (e.g. if the text is `one[] two`, with `[]` being the cursor, it tries to complete `one`, not `two`).

magik6k commented on src/main/resources/assets/opencomputers/loot/OpenOS/lib/term.lua:
```diff
@@ -247,29 +250,61 @@ function term.read(history, dobreak)
+        local h = hint(line())
+        if type(h) == "string" then
+          local _, cby = getCursor()
+          history[cby] = after
```
Ouch, this should be `history[cby] = h`

fnuecke commented:
Thanks again! Now to find some people to break it... :-P
